File: CommonLib/VirEnv/DataLogger.py
"""DataLogger -- CSV time-series logging of the FIXS vehicle-data stream.

Python peer of ``CommonLib/DataLogger.{h,cpp}``: same method names, same header
line, same ``%.3f`` / ``%.5f`` cell formatting, so a CSV written by the Python
bridge and one written by the C++ bridge are diffable line for line.

IMPORTANT -- the logged values are in the FIXS WIRE convention, which is the
SUMO / VISSIM convention:

* position = front-of-vehicle anchor, metres
* heading  = navigational degrees, 0 = North, clockwise
* speed    = m/s

The traffic simulators share that convention; the vehicle-dynamics / virtual-
environment backends (CARLA, CarMaker) each have their own native frames and
diverge more. Those are normalised to the wire convention by the bridge BEFORE a
record reaches this logger -- which is the whole reason to log at the FIXS-message
layer rather than at a backend-native transform: logs from ANY backend are then
directly comparable.

Analysis and plotting of a logger's output is TEST-SPECIFIC and lives under the
relevant ``tests/`` folder, never in this class.
"""

import os

__all__ = ['DataLogger']

#: Written when DataLogSetup names no fields. Same list as the C++ default.
_DEFAULT_FIELDS = ('positionX', 'positionY', 'positionZ', 'heading', 'speed')

#: Fields rendered as text rather than as a number.
_STRING_FIELDS = frozenset({'type', 'vehicleClass', 'linkId'})

#: Fields this logger knows how to render. An unknown name yields an empty cell
#: rather than an error, so a config naming a field this build does not have
#: still produces a usable log -- exactly as the C++ ``cell()`` does.
_NUMERIC_FIELDS = frozenset({
    'positionX', 'positionY', 'positionZ', 'heading', 'speed', 'speedDesired',
    'acceleration', 'grade', 'length', 'width', 'height',
})


class DataLogger:
    """A generic, config-driven CSV logger for FIXS vehicle records."""

    def __init__(self):
        self._f = None
        self._path = ''
        self._fields = list(_DEFAULT_FIELDS)

    def open(self, path, fields):
        """(string, list) -> bool -- open the CSV; False if it cannot be written.

        Empty ``fields`` selects the default core set. Parent directories are
        created as needed. A False return leaves the logger a no-op, so callers
        need no extra guard around :meth:`logVehicle`.
        """
        self._path = path
        self._fields = list(fields) if fields else list(_DEFAULT_FIELDS)

        parent = os.path.dirname(os.path.abspath(path))
        try:
            os.makedirs(parent, exist_ok=True)
            self._f = open(path, 'w', encoding='utf-8', newline='')
        except OSError:
            self._f = None
            return False

        # Self-documenting convention line; CSV readers skip lines starting '#'.
        self._f.write('# FIXS DataLogger -- FIXS wire convention (== SUMO/VISSIM): '
                      'position = front-of-vehicle anchor [m], heading = navigational '
                      'degrees (0=N, CW), speed [m/s]. Backend-native frames '
                      '(Carla/CarMaker) normalized upstream.\n')
        self._f.write('simTime,id' + ''.join(',' + f for f in self._fields) + '\n')
        self._f.flush()
        return True

    def logVehicle(self, simTime, v, idOverride=None):
        """(float, VehData[, string]) -> None -- append one record.

        The record's values must already be in the FIXS wire convention documented
        above.

        ``idOverride`` labels the row with a different id than the record
        carries, for logging two VIEWS of one vehicle on the same clock --
        the CARLA ego and the traffic simulator's copy of it, as 'ego' and
        'ego_sumo'. It belongs here rather than at the call site because a
        fixs record refuses assignment to `id` (it is measured data), and a
        copy carries the same guard.
        """
        if self._f is None:
            return
        self._f.write('%.3f,%s%s\n'
                      % (simTime, idOverride or v.id,
                         ''.join(',' + self._cell(f, v) for f in self._fields)))
        # Flushed per record: at a 10 Hz feed that is cheap, and it means the log
        # survives a hard kill of the demo -- which is when it is most wanted.
        self._f.flush()

    def isOpen(self):
        return self._f is not None

    def path(self):
        return self._path

    def close(self):
        if self._f is not None:
            self._f.close()
            self._f = None

    @staticmethod
    def _cell(f, v):
        """Format one field by name. Unknown -> an empty cell, as the C++ does."""
        if f in _NUMERIC_FIELDS:
            return '%.5f' % getattr(v, f, 0.0)
        if f in _STRING_FIELDS:
            return str(getattr(v, f, ''))
        return ''

File: CommonLib/DataLogger.cpp
//============================================================================
//  DataLogger -- CSV time-series logging of the FIXS vehicle-data stream.
//  See DataLogger.h for the (SUMO/VISSIM) coordinate/heading convention.
//============================================================================
#include "DataLogger.h"

#include <windows.h>     // CreateDirectoryA (CommonLib stays C++14: no std::filesystem)
#include <cstdio>

namespace fixs {

// Best-effort: create every ancestor directory of 'path' (Win32, C++14-safe).
static void ensureParentDir(const std::string& path) {
    std::string cur;
    for (char c : path) {
        if ((c == '/' || c == '\\') && !cur.empty() && cur.back() != ':')
            CreateDirectoryA(cur.c_str(), NULL);
        cur.push_back(c);
    }
}

bool DataLogger::open(const std::string& path, const std::vector<std::string>& fields) {
    path_   = path;
    fields_ = fields;
    if (fields_.empty())
        fields_ = { "positionX", "positionY", "positionZ", "heading", "speed" };

    ensureParentDir(path_);
    ofs_.open(path_, std::ios::out | std::ios::trunc);
    if (!ofs_.is_open()) return false;

    // Self-documenting convention line; CSV readers skip lines starting with '#'.
    ofs_ << "# FIXS DataLogger -- FIXS wire convention (== SUMO/VISSIM): position = "
            "front-of-vehicle anchor [m], heading = navigational degrees (0=N, CW), "
            "speed [m/s]. Backend-native frames (Carla/CarMaker) normalized upstream.\n";
    ofs_ << "simTime,id";
    for (const std::string& f : fields_) ofs_ << "," << f;
    ofs_ << "\n";
    ofs_.flush();
    return true;
}

void DataLogger::logVehicle(double simTime, const VehFullData_t& v) {
    if (!ofs_.is_open()) return;
    char t[32];
    std::snprintf(t, sizeof(t), "%.3f", simTime);
    ofs_ << t << "," << v.id;
    for (const std::string& f : fields_) ofs_ << "," << cell(f, v);
    ofs_ << "\n";
    ofs_.flush();   // 10 Hz feed rate -> cheap; survives a hard kill of the demo
}

std::string DataLogger::cell(const std::string& f, const VehFullData_t& v) {
    char b[48];
    auto num = [&](double x) { std::snprintf(b, sizeof(b), "%.5f", x); return std::string(b); };
    if (f == "positionX")     return num(v.positionX);
    if (f == "positionY")     return num(v.positionY);
    if (f == "positionZ")     return num(v.positionZ);
    if (f == "heading")       return num(v.heading);
    if (f == "speed")         return num(v.speed);
    if (f == "speedDesired")  return num(v.speedDesired);
    if (f == "acceleration")  return num(v.acceleration);
    if (f == "grade")         return num(v.grade);
    if (f == "length")        return num(v.length);
    if (f == "width")         return num(v.width);
    if (f == "height")        return num(v.height);
    if (f == "type")          return v.type;
    if (f == "vehicleClass")  return v.vehicleClass;
    if (f == "linkId")        return v.linkId;
    // The actuation channel. Logging these is how a run separates "was the
    // vehicle told to move?" from "did it move?" -- indistinguishable in a
    // position/speed trace, and with completely different causes.
    if (f == "acceleratorPedalDesired") return num(v.acceleratorPedalDesired);
    if (f == "brakePedalDesired")       return num(v.brakePedalDesired);
    if (f == "steerAngleDesired")       return num(v.steerAngleDesired);
    return "";                // unknown field -> empty cell (safe)
}

void DataLogger::close() {
    if (ofs_.is_open()) ofs_.close();
}

} // namespace fixs
